File: apps/auth-server/src/main/java/com/example/chat/auth/server/core/service/CompositeAuthService.java
package com.example.chat.auth.server.core.service;

import com.example.chat.auth.server.core.domain.AuthLevel;
import com.example.chat.auth.server.core.domain.AuthResult;
import com.example.chat.auth.server.core.domain.CredentialType;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.Set;

/**
 * 복합 인증 서비스
 */
@Service
public class CompositeAuthService {

    /**
     * 여러 자격증명 결과를 조합하여 최종 인증 수준 결정
     */
    public AuthResult combineResults(AuthResult first, AuthResult second) {
        if (!first.authenticated() || !second.authenticated()) {
            return AuthResult.failure("One or more authentication methods failed");
        }

        Set<CredentialType> combined = new HashSet<>(first.completedCredentials());
        combined.addAll(second.completedCredentials());

        AuthLevel finalLevel = first.authLevel().getLevel() > second.authLevel().getLevel()
                ? first.authLevel()
                : second.authLevel();

        return AuthResult.success(finalLevel, combined);
    }

    /**
     * 인증 수준 업그레이드
     */
    public AuthLevel upgradeAuthLevel(AuthLevel current, AuthResult additional) {
        if (additional.authLevel() != null && additional.authLevel().getLevel() > current.getLevel()) {
            return additional.authLevel();
        }
        return current;
    }
}

File: apps/auth-server/src/main/kotlin/com/example/chat/auth/server/core/service/CompositeAuthService.kt

package com.example.chat.auth.server.core.service

import com.example.chat.auth.server.core.domain.AuthLevel
import com.example.chat.auth.server.core.domain.AuthResult
import org.springframework.stereotype.Service

/**
 * 복합 인증 서비스
 */
@Service
class CompositeAuthService {

    /**
     * 여러 자격증명 결과를 조합하여 최종 인증 수준 결정
     */
    fun combineResults(first: AuthResult, second: AuthResult): AuthResult {
        if (!first.isAuthenticated || !second.isAuthenticated) {
            return AuthResult.failure("One or more authentication methods failed")
        }

        val combined = first.completedCredentials + second.completedCredentials

        val finalLevel = if (first.authLevel!!.level > second.authLevel!!.level) {
            first.authLevel
        } else {
            second.authLevel
        }

        return AuthResult.success(finalLevel, combined)
    }

    /**
     * 인증 수준 업그레이드
     */
    fun upgradeAuthLevel(current: AuthLevel, additional: AuthResult): AuthLevel {
        return if (additional.authLevel!!.level > current.level) {
            additional.authLevel
        } else {
            current
        }
    }
}
